pkg/controller/nstemplatetier/nstemplatetier_controller.go

```diff
@@ -27,100 +29,164 @@ import (
+	return &NSTemplateTierReconciler{
+		client: cl,
+		scheme: s,
+	}
+}
+
+// NSTemplateTierReconciler reconciles a NSTemplateTier object (only when this latter's specs were updated)
+type NSTemplateTierReconciler struct {
 	// This client, initialized using mgr.Client() above, is a split client
 	// that reads objects from the cache and writes to the apiserver
 	client client.Client
 	scheme *runtime.Scheme
 }
 
-// Reconcile reads that state of the cluster for a NSTemplateTier object and makes changes based on the state read
-// and what is in the NSTemplateTier.Spec
-// Note:
-// The Controller will requeue the Request to be processed again if the returned error is non-nil or
-// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcileNSTemplateTier) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+const (
+	// MaxPoolSize the maximum number of TemplateUpdateRequest resources that can exist at the same time
+	MaxPoolSize = 5
+)
+
+// Reconcile takes care of:
+// - inserting a new entry in the `status.updates` (and cleaning the 'failedAccounts` in the previous one)
+// - creating and delete the TemplateUpdateRequest to update the MasterUserRecord associated with this tier
+// - updating the `Failed` counter in the `status.updates` when a MasterUserRecord failed to update
+// - setting the `completionTime` when all MasterUserRecord have been processed
+func (r *NSTemplateTierReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	logger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
-	logger.Info("Reconciling NSTemplateTier")
 
-	// Fetch the NSTemplateTier tier
+	// fetch the NSTemplateTier tier
 	tier := &toolchainv1alpha1.NSTemplateTier{}
-	err := r.client.Get(context.TODO(), request.NamespacedName, tier)
-	if err != nil {
+	if err := r.client.Get(context.TODO(), request.NamespacedName, tier); err != nil {
 		if errors.IsNotFound(err) {
-			log.Info("NSTemplateTier not found")
+			logger.Info("NSTemplateTier not found")
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		logger.Error(err, "unable to get the current NSTemplateTier tier")
 		return reconcile.Result{}, err
 	}
-	if err := r.ensureTemplateUpdateRequest(logger, tier); err != nil {
+
+	// create a new entry in the `status.history`
+	if added, err := r.ensureStatusUpdateRecord(logger, tier); err != nil {
+		logger.Error(err, "unable to insert a new entry in status.updates after NSTemplateTier changed.")
+		return reconcile.Result{}, err
+	} else if added {
+		logger.Info("Requeing after adding a new entry in tier.status.updates")
+		return reconcile.Result{Requeue: true}, nil
+	}
+	if done, err := r.ensureTemplateUpdateRequest(logger, tier); err != nil {
 		logger.Error(err, "unable to ensure TemplateRequestUpdate resource after NSTemplateTier changed.")
```

Comment:
Let's wrap this error with NotReady status. Similar to https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/masteruserrecord/masteruserrecord_controller.go#L214:37
We usually use `wrapErrorWithStatusUpdate()` function to:
- log the error
- Add Ready=False condition with the original error message to the resource status.

Reply:
Yes, I've seen this `wrapErrorWithStatusUpdate()` func being used in other controller, I don't really understand why the `NSTemplateTier` resource should be in a `NotReady` condition in this case. 
In other words, what does it mean for the The `NSTemplateTier` resource to be `Ready` or `NotReady`?

Reply:
The point here is to have something in the status where we can see the error. Otherwise we will have to look up the errors in logs. It's not that convenient. We could use already available `Ready` condition here which could mean that the template tier is not ready to update templates. But we can introduce a new condition for that. `UpdateFailure` for example.
So, we use it for reporting errors and clean it up in the end of reconcile if everything is fine.

Reply:
ok, I get your point, but for that specific topic, may I suggest that we discuss about it when reviewing https://github.com/codeready-toolchain/host-operator/pull/243 for https://issues.redhat.com/browse/CRT-708?

Reply:
Yes, I'm fine with addressing it in https://github.com/codeready-toolchain/host-operator/pull/243

Reply:
thanks ;)